### ISSUE TITLE
Separate `lxc snapshot` command

### ISSUE THREAD

**Issue by rkelleyrtp** [lxc/lxd#3251]
Hoping this could get filed as a feature request.

Currently, the lxd snapshot feature (LXD 2.13) is broken into separate (seemingly disconnected) commands:
* lxc snapshot (to create the snapshot)
* lxc restore (to restore the snapshot)
* lxc copy <container>/<snapshot> <new_container>
* lxc info to list the snapshot
* ...

In my opinion, it is more intuitive to have all snapshot functionality live under the "snapshot" command.
Example:
* lxc snapshot list
* lxc snapshot create
* lxc snapshot delete
* lxc snapshot restore
* lxc snapshot copy

**Comment by stgraber** [comment]
Moving this to later and tagged as Maybe since doing so would break anything using the existing commands and so could only be done at the time of LXD 3.0 should we decide to break CLI stability then.

**Comment by SoulSeekkor** [comment]
Agreed, also the lxc delete to remove snapshots is extremely dangerous.  Even if it is left alone and still usable additional aliases like lxc snapshot delete etc would be appreciated.

**Comment by stgraber** [comment]
So obviously we missed our opportunity to do this for 3.0 but it's something I'm certainly considering for 4.0

**Comment by srados** [comment]
maybe I could look into this one

**Comment by stgraber** [comment]
@srados we're not going to touch this one until February/March right before we tag `4.0` as we can't include this work without completely breaking our CLI and we don't do such breaks outside of major version bumps.

**Comment by srados** [comment]
yes, I understand that. What I meant is having a separate branch started for v4.
Another solution would be keeping those changes backwards compatible, meaning, the old commands will still work. So it boils down to extending the commands parser and mapping it to already existing snapshot actions. I haven't checked the code yet, so I'm not sure if that's possible, but I suppose it is.

**Comment by stgraber** [comment]
So the problem is that right now `lxc snapshot list` for example is a perfectly valid command if you want to create an automatically named snapshot for a container called `list`.

That's why we can't really do any of this work in a backward compatible way. We know we're going to break people as soon as we merge this and so intend to do it at the last minute.

You could certainly prepare a branch already which adds the new subcommand and deprecates the old commands but there's a pretty good chance that we'll be doing some kind of change to the current snapshot handling functions between now and the 4.0 release which would then require changes to that branch.

Given the work itself shouldn't take more than a few hours, it feels it may be better to write that code once we're ready to merge such breaking changes.

**Comment by srados** [comment]
oh, I get it. OK, I guess, then backwards compatibility is impossible (or hard, or impractical). I agree, let's wait until the 4.0 approaches

**Comment by stgraber** [comment]
So I implemented this pretty quickly, it's not a very hard change.
However just looking at the amount of tests and examples we'd have to update within the LXD codebase and our website alone, this will be a very very disruptive change.

We've also implemented the same CLI patterns in other places (lxc storage) so we'd need to change and break those too.

Overall, I agree that snapshots should be separate, but consistency in this case comes at far too great a cost to inflict it on our users. We will certainly not be making the same design decision when comes the time to rewrite the CLI entirely though.
